### ISSUE TITLE
5.0.6 - Invoke-DbaQuery with Query "UPDATE" does not work!

### ISSUE THREAD

**Issue by donlent** [ironmansoftware/powershell-universal#3759]
### Version

5.0.6

### Severity

Critical

### Environment

msi

### Steps to Reproduce

Unfortunately another showstopper found with a core functionality... Sorry Adam... :-(

`Invoke-DbaQuery -SqlInstance $mySQLInstance -Query "UPDATE MyTable SET Name = 'My New Name', Description = 'My New Description' WHERE Id = $EventData.Id"`

So far SELECT and INSERT works. What's wrong with UPDATE??

BTW executing this exact command from outside PSU works! So it's sort of a compatibility Issue with PSU and dbatools and this Query parameter maybe??

How are you folks doing it?

Thanks!!

### Expected behavior

```console
UPDATE should work
```

### Actual behavior

```console
UPDATE does not work
```

**Comment by adamdriscoll** [comment]
Is there an error or does it just not do anything?

**Comment by donlent** [comment]
No error at all. It just does nothing.

**Comment by adamdriscoll** [comment]
What environment are you running in (integrated vs ps7)? I'll give it a go.

**Comment by donlent** [comment]
ps7.. thanks a lot! really sorry about that! :-(

**Comment by MikeShepard** [comment]
Should it be $($Eventdata.id) ?

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by donlent** [comment]
Thank you so much @MikeShepard but indeed, I tried that as well. Also by statically putting the Id... No way. It's really just when fired-up from within the PSU app. outside the exact same code works.
So strange...

**Comment by adamdriscoll** [comment]
This works for me. I'm using dbatools 2.1.23 I would recommend trying to run the command in a script and use the -Verbose parameter to get some more information. 

When you run the script, make sure you set the verbose preference, so the verbose messages show up.

![image](https://github.com/user-attachments/assets/2b6204c8-04de-4984-9f8f-09c0f920fab4)

This is my test script. 

```powershell
$server = Connect-DbaInstance -SqlInstance "(localdb)\mssqllocaldb" -Database PSU
Invoke-DbaQuery -SqlInstance $server -Query "UPDATE Computer SET Name = 'TestComputer123' WHERE Id = 2" -Verbose
```

This is the output. 

```
Populating RepositorySourceLocation property for module dbatools.library.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\dbatools.library.psd1'.
Populating RepositorySourceLocation property for module dbatools.library.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\dbatools.library.psm1'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\win-sqlclient\Microsoft.Data.SqlClient.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\third-party\bogus\bogus.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Server.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Dac.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Smo.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.SmoExtended.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.SqlWmiManagement.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.WmiEnum.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Management.RegisteredServers.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Management.Collector.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Management.XEvent.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.Management.XEventDbScoped.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.SqlServer.XEvent.XELite.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\third-party\LumenWorks\LumenWorks.Framework.IO.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Azure.Core.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Azure.Identity.dll'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\Microsoft.IdentityModel.Abstractions.dll'.
Exporting function 'Get-DbatoolsLibraryPath'.
Importing function 'Get-DbatoolsLibraryPath'.
Loading module from path 'C:\Users\adamr\OneDrive\Documents\PowerShell\Modules\dbatools.library\2024.4.12\core\lib\dbatools.dll'.
Importing cmdlet 'Import-Command'.
Importing cmdlet 'Read-XEvent'.
Importing cmdlet 'Select-DbaObject'.
Importing cmdlet 'Set-DbatoolsConfig'.
Importing cmdlet 'Write-Message'.
[11:19:01][Register-DbaRunspace] Registering runspace: dbatools-logging
[11:19:01][Start-DbaRunspace] Starting runspace: dbatools-logging
[11:19:01][Register-DbaRunspace] Registering runspace: dbatools-teppasynccache
[11:19:01][Start-DbaRunspace] Starting runspace: dbatools-teppasynccache
[11:19:01][Register-DbaRunspace] Registering runspace: dbatools-maintenance
[11:19:01][Start-DbaRunspace] Starting runspace: dbatools-maintenance
[11:19:02][Connect-DbaInstance] Starting loop for '(localdb)\mssqllocaldb': ComputerName = 'localhost', InstanceName = 'mssqllocaldb', IsLocalHost = 'True', Type = 'Default'
[11:19:02][Connect-DbaInstance] String is passed in, will build server object from instance object and other parameters, do some checks and then return the server object
[11:19:02][Connect-DbaInstance] authentication method is 'local integrated'
[11:19:02][Connect-DbaInstance] localdb detected, skipping unsupported keyword 'Encryption'
```

You can see the database was updated. 

![image](https://github.com/user-attachments/assets/6f24fa76-ed1d-4c57-9f62-9b5c5850adc6)

**Comment by adamdriscoll** [comment]
Lol sorry about those images...didn't realize they would blow up like that.

**Comment by donlent** [comment]
Thanks a lot Adam! So strange! Will give it a try hopefully tomorrow or latest on Monday using exactly what you've posted.

And don't worry about the images. I have the same issue with the big screen when I take screenshots... :-) 

Will report back as soon as I've tested it.

THANKS AGAIN!!

**Comment by donlent** [comment]
Hello @adamdriscoll 
As promised I took some time to dig deeper into the issue and I think I found the source of the problem!

First info ahead, using the script method as mentioned from you and testing from there it worked as expected. So Why didn't it work from the App itself?

It came out that it was not related to the UPDATE itself but apparently any Invoke-DbaQuery execution **_does not work if executed from within an -OnRender ScriptBlock!_**

Example of NON-WORKING Function:
```
New-UDTableColumn -Property Action -Title "Action" -OnRender {
        New-UDButton -Text "TestSave1" -OnClick {
            Invoke-DbaQuery -SqlInstance $myInstance -Query "UPDATE Customers SET Name = 'Hello Customer FromOnRender', Description = 'Hello Description FromOnRender' WHERE Id = 3" -Verbose
            Write-Warning "Invoke-DbaQuery executed but no changes perfomed in the DB!"
        }
```

Now if you put the exact same code outside the -OnRender so let's say you put it in the ToolbarContent of the Table, guess what it works!

Example of WORKING Function:
```
New-UDTable -Icon (New-UDIcon -Icon "database") -Title " Manage Customers" -Data $Customers -Columns $Columns -ShowSearch -DisableSortRemove -ShowSort -ShowFilter -ShowPagination -ToolbarContent {

    # Add New Customer
    New-UDButton -Text "TestSave2" -OnClick {
Invoke-DbaQuery -SqlInstance $myInstance -Query "UPDATE Customers SET Name = 'Hello Customer FromOnRender', Description = 'Hello Description FromOnRender' WHERE Id = 3" -Verbose
            Write-Warning "Invoke-DbaQuery executed and changes are there!"
        }
}
```

So it seems that it is definitely related to something with OnRender which (at least) dbatools doesn't like.

Hope you can find out what it is, since this is really a big issue right now...

Thanks a lot in advance for your help and I wish you a very good start in the new week ahead!

**Comment by adamdriscoll** [comment]
I have no idea why there would be a difference, but I'll see if I can reproduce. One thing you could consider doing is running the function outside of the app in a script to avoid any kind of weird assembly binding issues. 

Script

```powershell
param($CustomerId)

            Invoke-DbaQuery -SqlInstance $myInstance -Query "UPDATE Customers SET Name = 'Hello Customer FromOnRender', Description = 'Hello Description FromOnRender' WHERE Id = $CustomerId" -Verbose
```

App

```powershell
New-UDTableColumn -Property Action -Title "Action" -OnRender {
        New-UDButton -Text "TestSave1" -OnClick {
            Invoke-PSUScript -Name 'UpdateDb.ps1' -Environment 'PowerShell 7' -Id $CustomerId -Wait | Out-Null
        }
```

**Comment by donlent** [comment]
Hey @adamdriscoll 
OK thank you so much, will try that and report back, hopefully already today.

Is that a general best-practice or just for this particular issue in order to try to overcome the issue? Because I've never done it like that before... So just wanted to learn what is your recomendation regarding that.

Thanks,
Don

**Comment by adamdriscoll** [comment]
I'd say it's best practice in the sense that you have an audit log of modifications made to the database (job log) if you run scripts rather than just perform the actions directly against the database. That said, it really depends on your purpose. I was just suggesting this because it seems to be something environmental and it helps to isolate the dbatools module from a bunch of things that PSU loads. 

That said, I tried to reproduce this myself and I still don't have an issue updating the database. This was my test app. 

```
New-UDDashboard -Title "DB" -Content {
$Data = @(
    @{Dessert = 'Frozen yoghurt'; Calories = 159; Fat = 6.0; Carbs = 24; Protein = 4.0; }
    @{Dessert = 'Ice cream sandwich'; Calories = 159; Fat = 6.0; Carbs = 24; Protein = 4.0}
    @{Dessert = 'Eclair'; Calories = 159; Fat = 6.0; Carbs = 24; Protein = 4.0}
    @{Dessert = 'Cupcake'; Calories = 159; Fat = 6.0; Carbs = 24; Protein = 4.0}
    @{Dessert = 'Gingerbread'; Calories = 159; Fat = 6.0; Carbs = 24; Protein = 4.0}
) 

New-UDTable -Data $Data -ShowSearch -Columns @(
    New-UDTableColumn -Property 'Dessert'
    New-UDTableColumn -Property 'Calories'
    New-UDTableColumn -Property Action -Title "Action" -OnRender {
        New-UDButton -Text "TestSave1" -OnClick {
            $server = Connect-DbaInstance -SqlInstance "(localdb)\mssqllocaldb" -Database PSU
            Invoke-DbaQuery -SqlInstance $server -Query "UPDATE Computer SET Name = 'TestComputer123' WHERE Id = 2" -Verbose
        }
    }
) 
}

```

**Comment by donlent** [comment]
Hey Adam
Thanks a lot. That is sooo strange, really! :-(

In the meantime I tried also with your suggestion, however I don't get it to run because this Cmdlet (Invoke-PSUScript) doesn't honor the -TrustCertificate Switch for whatever reason. Even if it's specified also in the appsettings.json, that doesn't work at all.
I always get this error:
![image](https://github.com/user-attachments/assets/d5cba987-bafe-4727-8d9e-06394df3f98e)

So atm not really possible to try even with your suggestion.

Just note, I'm still using 5.0.6. According to 5.0.7 release notes you state this:
Fixed an issue with TrustCertificate in appsettings.json

So maybe this issue will be resolved there? Unfortunately I need to stop here for today. But I'll update the environment tomorrow morning and retry.

...I'll also try to figure out what's happening with this OnRender issue and report back.

Again, thanks a lot for your much appreciated help!🙏

**Comment by donlent** [comment]
Hello @adamdriscoll 
Sorry if I didn't manage to get back to you yesterday.
In the meantime I was able to finally find the root cause to the issue! :-D

It is related to scoping! Since I had my connection definition at the very top of the page, like that it seems to work almost everywhere but not inside a -OnRender Scriptblock.

So I managed to workaround the issue by putting the definition inside a Session Variable.
Like that it works everywhere, no matter where I call the functions.

I will close the issue therefore. In case you have the chance to answer to this comment is this by design or is it really a "bug" and a Variable defined as non-session-scoped should work as well?

Again, thanks a lot for your much appreciated help and commitment! Your support is really as always awesome! All product developers should take you as an example! 🙏🤗

Best regards,
Don
